### PR TITLE
Pre2P - RainTree

### LINKED ISSUE #90: [P2P] Pre2P RainTree
# Objective
Implement the RainTree algorithm described [here]() on top of the Pre2P module.

# Origina Document
This issue was created after the work in https://github.com/pokt-network/pocket/pull/80 was mostly complete but created for completeness.

# Goals
## Deliverables
- [X] Implementaiton
- [X] Tests
- [X] Documentaiton

## [Optional] Testing Methodology
- [X] Add new unit tests
- [X] Create a new test suite for running RainTree tests (similar to HotPOKT)
- [X] Verify that everything works on the 4 node LocalNet framework described [here](https://github.com/pokt-network/pocket/tree/main/docs/development)

## Non-goals
- Migrate Pre2P to P2P (to be done in a different issue)
- Implement all the components of a fully functional P2P layer (peer discovery, redundancy, complete configs, async IO, etc...)

Creator: @Olshansk 
Co-Owners: @andrewnguyen22 @derrandz

### PR DESCRIPTION
**Pull request by Olshansk**
Pocket Network's RainTree implementation as described [here](https://github.com/pokt-network/pocket-network-protocol/tree/main/p2p) atop of the Pre2P module.

# Changes

## General
- Updated Makefile to generate mocks for the bus & network interfaces
- Improve leader election logging
- Added a `ValidatorJsonCompatibleWrapper` because the protobuf structure cannot be imported from a json Genesis file

## Node Configs
- Add `UseRainTree` and `ConnectionType` options to `Pre2PConfig`
- Reordered the private keys in the configs to make ordering easier to understand in development
- Modified LocalNet to use `tcp` connection and `RainTree` by default
- Explicitly specify validators in the LocalNet config

## Development Configs
- Modified `node.consensus` node to enable Delve debugging (tested with VSCode but not Goland)

## Pre2P
- Moved `pre2p/handlers.go` in place to reduce # files
- Abstracted out TCP specific functionality in `pre2p` to a general `TransportLayerConn` to enable injecting an override for testing purposes
- Extract the previous approach to networking into `stdnetwork`, which is replaced by `raintree` but can be toggled back
- Move validator related helpers into their own utils file

## RainTree
- Partial implementation of RainTree implementing the `Network` interface, which is a substitute for the previous `std network`
    - Just the end-to-end “happy path”
- End-to-end RainTree tests of the P2P module in isolation for a 4 node network and a 9 node network

# Testing

## Pre2P tests

```
$ make test_pre2p
```

## All tests

```
$ make test_all
```

## LocalNet

Follow the instructions in [docs/development/README.md](./docs/development/README.md)

# TODOs

## TODO - After review
- [ ] Test that Delve debugging also works with Goland

## TODO - Did not finish
- [ ] RainTree features - need to implement and add tests
    - [ ] ACK/Adjust/Resend 
    - [ ] Daisy chain cleanup (i.e. IGYW messages)
    - [ ] Add unit tests for partial visibility
    - [ ] Add unit tests for dead/faulty nodes (both deterministic and ephemeral)
- [x] Change the `raintTreeNetwork` specific methods in `addrbook_utils` to `functions` that are not tied to a struct instance
- [x] Unit tests for `addrbook_utils.go`
- [x] Move logs and errors into a shared file

## TODO - Future commits
- [ ] Completely refactor `typesGenesis.GetNodeState` which is a poor overall design
- [ ] Various `TODO` types left throughout the code
- [ ] Delete `stdnetwork` altogether and simplify code by embedding `Braintree` in place
- [ ] Improve logging & telemetry 
- [ ] Look into `uber-go/dig` or `uber-go/fx` to use a proper dependency injection network

### CODE REVIEW

**Comment by Olshansk** [comment]
@andrewnguyen22 I reviewed at a high level (not in depth). It makes sense, but I'd prefer to separate it out into a subsequent PR where I will add more tests, and merge Pre2P RainTree as is in the meantime.

**Comment by Olshansk** [comment]
@andrewnguyen22 I reviewed the original TODOs I left in my PR description and here is what I'm thinking of as next steps:

0. Finish the following TODOs in this Branch
- [ ] Change the `raintTreeNetwork` specific methods in `addrbook_utils` to `functions` that are not tied to a struct instance
- [ ] Unit tests for `addrbook_utils.go`
- [ ] Move logs and errors into a shared file

1. Create a PR issue with the `infra` tag for:
- [ ] Test that Delve debugging also works with Goland
- [ ] Look into `uber-go/dig` or `uber-go/fx` to use a proper dependency injection network throughout
- [ ] Improve logging & telemetry in Pre2P & P2P (cc @derrandz)

2. Create a "RainTree redundancy" PR issue and split out the commit you submitted in this branch in there with the following tasks:
- [ ] RainTree features & Tests
    - [ ] ACK/Adjust/Resend 
    - [ ] Daisy chain cleanup (i.e. IGYW messages)
    - [ ] Add unit tests for partial visibility
    - [ ] Add unit tests for dead/faulty nodes (both deterministic and ephemeral)
    - [ ] Add a list of TODOs in the code for more extensive tests we want to have in the future

3. Create PR issues with the `tech debt` tag for these issues:
- [ ] Completely refactor `typesGenesis.GetNodeState` which is a poor overall design
- [ ] Delete `stdnetwork` altogether and simplify code by embedding `Braintree` in place
- [ ] Various `TODO` types left throughout the code

This way, the only outstanding tasks to submit the original Pre2P RainTree implementation will be finishing the tasks in (0).

**Comment by Olshansk** [comment]
Remaining action items on Olshansky:
- [ ] Finish adding unit tests for `addrbook_utils.go`
- [ ] Move logs and errors into a shared file
- [ ] Add a `README.md` describing the code layout

Dependant issues:
- [ ] https://github.com/pokt-network/pocket/issues/85
- [ ] https://github.com/pokt-network/pocket/issues/86

**Comment by Olshansk** [comment]
@andrewnguyen22 I know you've already taken a look at the core implementation, so the two pieces where I'd like more review is:

1. Do you think `Move logs and errors into a shared file` is necessary?
2. Feedback on documentation at `pocket/p2p/pre2p/README.md`
3. Feedback on tests at `pocket/p2p/pre2p/raintree/addrbook_utils_test.go` and `/pocket/p2p/pre2p/raintree_utils_test.go`.

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]

**Comment by andrewnguyen22** [comment]
Fix makefile conflicts

**Comment by andrewnguyen22** [comment]
TLDR; really good contribution - solid tests, just a few nits and we can merge

File: p2p/pre2p/raintree/addrbook_utils_test.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+		{81, 4},
+		// 5 levels
+		{82, 5},
+		// 10 levels
+		{59049, 10},
+		// 11 levels
+		{59050, 11},
+		// 19 levels
+		// INVESTIGATE(olshansky/team): Does not scale to 1,000,000,000 nodes
+		// {1000000000, 19},
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/raintree/addrbook_utils_test.go:63]: I changed the comment to:

```
		// NOTE: This does not scale to 1,000,000,000 (1B) nodes because it's too slow.
		//       However, optimizing the code to handle 1B nodes would be a very premature optimization
		//       at this stage in the project's lifecycle, so the comment is simply left to inform
		//       future readers.
```

**Comment by Olshansk** [comment]
Re various tasks, done the following:
1. Fixed typos
2. Used _ *testing.T in all testing functions
3. ValId removed where unused
3. createConfigs updated
4. FIxed make confilits

Re structure:
1. I consolidated the unit tests and helpers into 1 file
2. I renamed it so that you can tell that they are unit tests at the module level. Not quite integration or E2E, but also higher than just a basic unit test. I tried using reflections and moving it around, but it ended up being too complex.
3. I updated the graph so you can see how we can eventually remove `stdnetwork` and that it is a dropin for `raintree`

```
├── raintree
│   ├── addrbook_utils.go        # AddrBook utilities
│   ├── addrbook_utils_test.go   # AddrBook utilities unit tests
│   ├── network.go               # Implementation of the Network interface using RainTree's specification
│   └── types
│       └── proto
│           └── raintree.proto
├── module_raintree_test.gogo            # RainTree unit tests at the module level
├── stdnetwork                  # This can eventually be deprecated once raintree is 
│   └── network.go              # Implementation of the Network interface using Golang's std networking lib
```

`2. This doesn't render for me - could be missing something`

Let me know if the following image helps clarify things
Screen Shot 2022-06-09 at 1 00 05 PM

File: p2p/pre2p/raintree/addrbook_utils.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	return n.getTarget(level, secondMsgTargetPercentage)
+}
+
+func (n *rainTreeNetwork) getTarget(level uint32, targetPercentage float64) (cryptoPocket.Address, bool) {
+	// OPTIMIZE(olshansky): We are computing this twice for each message, but it's not that expensive.
+	l := n.getAddrBookLengthAtHeight(level)
+
+	i := int(targetPercentage * float64(l))
+
+	// If the target is 0, it is a reference to self, which is a `Demote` in RainTree terms.
```
Comment (derrandz) [inline p2p/pre2p/raintree/addrbook_utils.go:70]: I think I especially like this part

File: p2p/pre2p/raintree/network.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	if err := n.handleAddrBookUpdates(); err != nil {
+		// DISCUSS(drewsky): if this errors, the node could still function but not participate in
+		// message propagation. Should we return an error or just log?
+		log.Println("[ERROR] Error initializing rainTreeNetwork: ", err)
+	}
+
+	return typesPre2P.Network(n)
+}
+
+func (n *rainTreeNetwork) NetworkBroadcast(data []byte) error {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:58]: What's the reason for the duplication in the name? I find `Broadcast` sufficient.

File: p2p/pre2p/raintree/network.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	return n.networkBroadcastInternal(data, n.maxNumLevels, getNonce())
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:59]: I think this would be better named as `broadcastAtLevel`

File: p2p/pre2p/raintree/network.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+		}
+	}
+
+	if addr2, ok := n.getSecondTargetAddr(level); ok {
+		if err = n.networkSendInternal(bz, addr2); err != nil {
+			log.Println("Error sending to peer during broadcast: ", err)
+		}
+	}
+
+	if err = n.demote(msg); err != nil {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:90]: My favorite part about this implementation is this!

File: p2p/pre2p/raintree/network.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+		log.Println("Error demoting self during RainTree message propagation: ", err)
+	}
+
+	return nil
+}
+
+func (n *rainTreeNetwork) demote(rainTreeMsg *typesPre2P.RainTreeMessage) error {
+	if rainTreeMsg.Level > 0 {
+		if err := n.networkBroadcastInternal(rainTreeMsg.Data, rainTreeMsg.Level-1, rainTreeMsg.Nonce); err != nil {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:99]: I think this is a good example of how `broadcastAtLevel` would be more semantic than the current name.

File: p2p/pre2p/raintree/network.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *rainTreeNetwork) NetworkSend(data []byte, address cryptoPocket.Address) error {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:106]: Same question about broadcast, can we do just `Send`?

File: p2p/pre2p/raintree/network.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+		Data:  data,
+		Nonce: getNonce(),
+	}
+
+	bz, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	return n.networkSendInternal(bz, address)
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:118]: I think `networkSendInternal` and `NetworkSend` can just become `Send`. I don't see `networkSendInternal` used anywhere else. Correct me if I am wrong.

File: p2p/pre2p/raintree/network.go [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+	networkMessage := types.PocketEvent{}
+	if err := proto.Unmarshal(rainTreeMsg.Data, &networkMessage); err != nil {
+		log.Println("Error decoding network message: ", err)
+		return nil, err
+	}
+
+	// Continue RainTree propagation
+	if rainTreeMsg.Level > 0 {
+		if err := n.networkBroadcastInternal(rainTreeMsg.Data, rainTreeMsg.Level-1, rainTreeMsg.Nonce); err != nil {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:154]: As above

File: p2p/pre2p/raintree/network.go [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+		}
+	}
+
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(rainTreeMsg.Nonce))
+	hash := cryptoPocket.SHA3Hash(b)
+	hashString := hex.EncodeToString(hash)
+	// Avoids this node from processing a messages / transactions is has already processed at the
+	// application layer. The logic above makes sure it is only propagated and returns.
+	if n.mempool.Contains(hashString) {
```
Comment (derrandz) [inline p2p/pre2p/raintree/network.go:165]: I think this is good enough de-dupping, wdyt?

File: p2p/pre2p/types/network.go [hunk 11]
```diff
@@ -1,18 +1,40 @@
...
+	Address    cryptoPocket.Address
+	ServiceUrl string // This is only included because it's a more human-friendly differentiator between peers
+}
+
+type TransportLayerConn interface {
+	IsListener() bool
+	Read() ([]byte, error)
+	Write([]byte) error
+	Close() error
 }
```
Comment (derrandz) [inline p2p/pre2p/types/network.go:40]:
```suggestion
type Transport interface {
	IsListener() bool
	IsDialer() bool
	Accept() (net.Conn, error)
	Dial(...) (net.Conn, error)
	Close() error
}
```

Comment (derrandz) [reply p2p/pre2p/types/network.go:40]: This is a suggestion, I think the name can definitely change, in terms of interface methods and behavior, I am not sure we can do it before #86 so we can keep it as is until then.

File: p2p/pre2p/utils.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+		}
+		book = append(book, networkPeer)
+	}
+	return book, nil
+}
+
+// CLEANUP(drewsky): These functions will turn into more of a "ActorToAddrBook" when we have a closer
+// integration with utility.
+func ValidatorToNetworkPeer(cfg *config.Pre2PConfig, v *typesGenesis.Validator) (*typesPre2P.NetworkPeer, error) {
+	conn, err := CreateDialer(cfg, v.ServiceUrl)
```
Comment (derrandz) [inline p2p/pre2p/utils.go:31]: This is something we should think about, how many open connections do we want to maintain. Referencing #54

File: README.md [hunk 13]
```diff
@@ -44,6 +45,8 @@ All the links you'll need are listed below. If you'd like to contribute to the P
 
 - [Consensus Changelog](consensus/CHANGELOG.md)
 - [Utility Changelog](utility/CHANGELOG.md)
+- _Coming Soon: P2P Changelog_
+- _Coming Soon: Persistence Changelog_
```
Comment (derrandz) [inline README.md:49]: 🤔 We should really make sure merges don't go without changing these files. Maybe time for some pre commit tooling to validate this?

File: build/config/config1.json [hunk 14]
```diff
@@ -1,9 +1,11 @@
 {
   "root_dir": "/go/src/github.com/pocket-network",
   "genesis": "build/config/genesis.json",
-  "private_key": "2b000000000000000000000000000000000000000000000000000000000000001ba66c6751506850ae0787244c69476b6d45fb857a914a5a0445a24253f7b810",
+  "private_key": "2e00000000000000000000000000000000000000000000000000000000000000264a0707979e0d6691f74b055429b5f318d39c2883bb509310b67424252e9ef2",
   "pre2p": {
-    "consensus_port": 8080
+    "consensus_port": 8080,
+    "use_raintree": true,
+    "connection_type": "tcp"
```
Comment (derrandz) [inline build/config/config1.json:8]: I don't think we should open this for configuration. I think we should stick to TCP for everything for the moment until and do UDP for redundancy, clean-up and churn. So really no point opening this up for config.

File: build/config/config2.json [hunk 15]
```diff
@@ -1,9 +1,11 @@
 {
   "root_dir": "/go/src/github.com/pocket-network",
   "genesis": "build/config/genesis.json",
-  "private_key": "2c00000000000000000000000000000000000000000000000000000000000000f868bcc508133899cc47b612e4f7d9d5dacc90ce1f28214a97b651baa00bf6e4",
+  "private_key": "2d00000000000000000000000000000000000000000000000000000000000000ee37d8c8e9cf42a34cfa75ff1141e2bc0ff2f37483f064dce47cb4d5e69db1d4",
   "pre2p": {
-    "consensus_port": 8080
+    "consensus_port": 8080,
+    "use_raintree": true,
+    "connection_type": "tcp"
```
Comment (derrandz) [inline build/config/config2.json:8]: Similarly the rest.

**Review by derrandz — CHANGES_REQUESTED** [review]
I feel slightly strongly about some names in here, let me know if you can make those changes.

Overall, here are my thoughts:

In terms of implementation correctness for the networking part, you are now very familiar with most of my comments and they will be addressed in #86 and #54 and some others perhaps, but I think this makes for a very good prototype overall, as the code remedies itself, so to say.

On the other hand, when it comes to abstraction, I think I was not at ease initially with the two abstractions `TransportLayerConn` and `Network`.

As I read through, I think it makes for a good base abstraction and separates things well, decoupling logic from networking, so kudos for that.

Tests were good to read through, my issue was that I had to go look for those prepare and test functions outside of the context of the test case, but you can't complain as that DRYs code well enough. Only that I think it would be more semantic to name things to signify important stages of the test: `Setup` `Run` `Teardown`.

The best part of this PR is the rain tree implementation, names are verbose but everything is very clear.

Good work on this @Olshansk I think this is heading a good direction.

File: p2p/pre2p/module.go [hunk 16]
```diff
@@ -67,12 +78,12 @@ func (m *p2pModule) Start() error {
...
 		for {
-			conn, err := m.listener.AcceptTCP()
+			data, err := m.listener.Read()
 			if err != nil {
-				log.Println("Error accepting connection: ", err)
+				log.Println("Error reading data from connection: ", err)
 				continue
 			}
-			go m.handleNetworkMessage(conn)
+			go m.handleNetworkMessage(data)
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module.go:86]: ACK

File: p2p/pre2p/conn.go [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+func (c *tcpConn) Write(data []byte) error {
+	if c.IsListener() {
+		return fmt.Errorf("connection is a listener")
+	}
+
+	client, err := net.DialTCP(TCPNetworkLayerProtocol, nil, c.address)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/conn.go:101]: ACK

File: p2p/pre2p/module.go [hunk 18]
```diff
@@ -67,12 +78,12 @@ func (m *p2pModule) Start() error {
 
 	go func() {
 		for {
-			conn, err := m.listener.AcceptTCP()
+			data, err := m.listener.Read()
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module.go:81]: ACK

File: p2p/pre2p/raintree/addrbook_utils.go [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+const (
+	firstMsgTargetPercentage  = float64(1) / float64(3)
+	secondMsgTargetPercentage = float64(2) / float64(3)
+	shrinkagePercentage       = float64(2) / float64(3)
+	maxLevelsLogBase          = float64(3)
+	floatPrecision            = float64(0.0000001)
+)
+
+// Whenever `addrBook` changes, we also need to update `addrBookMap` and `addrList`
+func (n *rainTreeNetwork) handleAddrBookUpdates() error {
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/raintree/addrbook_utils.go:24]: Done.

File: p2p/pre2p/raintree/network.go (hunk 5 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:90]: ACK.

File: p2p/pre2p/raintree/network.go (hunk 6 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:99]: DONE + AGGREGATE. Renaming `networkBroadcastInternal` to `networkBroadcastAtLevel`.

File: p2p/pre2p/raintree/network.go (hunk 4 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:59]: DONE + AGGREGATE

File: p2p/pre2p/raintree/network.go (hunk 7 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:106]: AGGREGATE

File: p2p/pre2p/raintree/network.go (hunk 8 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:118]: Used in multiple places:

Screen Shot 2022-06-15 at 6 49 29 PM

File: p2p/pre2p/raintree/network.go (hunk 9 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:154]: AGGREGATE

File: README.md (hunk 13 above)
Comment (Olshansk) [reply README.md:49]: We can just ping each other when we find this.

My answer would be sure as long as it doesn't detract from our day-to-day responsibilities. Even better, create a Github issue for it as a starter task for someone in the community.

File: build/config/config2.json (hunk 15 above)
Comment (Olshansk) [reply build/config/config2.json:8]: AGGREGATE

File: build/config/config1.json (hunk 14 above)
Comment (Olshansk) [reply build/config/config1.json:8]: I'm using it for testing:

```
			Pre2P: &config.Pre2PConfig{
				ConsensusPort:  8080,
				UseRainTree:    true,
				ConnectionType: config.EmptyConnection,
			},
```

This way we inject the fact that we don't need TCP or UDP for mocking purposes.

File: p2p/pre2p/utils.go (hunk 12 above)
Comment (Olshansk) [reply p2p/pre2p/utils.go:31]: ACK

File: p2p/pre2p/raintree/network.go (hunk 10 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:165]: Adding this TODO: 
```
// TODO(team): Add more tests to verify this is sufficient for deduping purposes.
```

File: p2p/pre2p/raintree/network.go (hunk 3 above)
Comment (Olshansk) [reply p2p/pre2p/raintree/network.go:58]: AGGREGATE

File: p2p/pre2p/module_raintree_test.go [hunk 20]
```diff
@@ -0,0 +1,392 @@
...
+	go func() {
+		messageHandeledWaitGroup.Wait()
+		close(done)
+	}()
+
+	// Timeout or succeed
+	select {
+	case <-done:
+	// All done!
+	case <-time.After(3 * time.Second):
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module_raintree_test.go:228]: Yes. No race conditions.

File: p2p/pre2p/types/network.go (hunk 11 above)
Comment (Olshansk) [reply p2p/pre2p/types/network.go:40]: 1. I'm open to changing `TransportLayerConn` to `Transport`, but can you get a 2nd opinion from someone or a reference showing it's the way to go?

2. Do you think it's a good idea to start adding unused interface functions here before we have a need for them?

File: p2p/pre2p/module_raintree_test.go [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+		validatorId(t, 25): {1, 1},
+		validatorId(t, 26): {1, 1},
+		validatorId(t, 27): {1, 1},
+	}
+	testRainTreeCalls(t, originatorNode, expectedCalls, false)
+}
+
+// ### RainTree Unit Helpers - To remove redundancy of code in the unit tests ###
+
+func testRainTreeCalls(t *testing.T, origNode string, testCommConfig TestRainTreeCommConfig, isOriginatorPinged bool) {
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module_raintree_test.go:179]: I already left comments through such as `// Network configurations`, `// Test configurations`, `// Network initialization`.

Please be more specific.

File: p2p/pre2p/module.go [hunk 22]
```diff
@@ -25,22 +25,33 @@ var _ modules.P2PModule = &p2pModule{}
 type p2pModule struct {
 	bus modules.Bus
 
-	listener *net.TCPListener
-	network  typesPre2P.Network
+	listener typesPre2P.TransportLayerConn
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module.go:28]: It's meant to be a general purpose transport layer. Not TCP.

AGGREGATE + REJECT>

File: p2p/pre2p/module.go [hunk 23]
```diff
@@ -113,6 +124,30 @@ func (m *p2pModule) Send(addr cryptoPocket.Address, msg *anypb.Any, topic types.
 	return m.network.NetworkSend(data, addr)
 }
 
-func (m *p2pModule) GetAddrBook() []*typesPre2P.NetworkPeer {
-	return m.network.GetAddrBook()
+func (m *p2pModule) handleNetworkMessage(networkMsgData []byte) {
+	appMsgData, err := m.network.HandleNetworkData(networkMsgData)
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply p2p/pre2p/module.go:128]: See the interface

```
type Network interface {
	NetworkBroadcast(data []byte) error
	NetworkSend(data []byte, address cryptoPocket.Address) error

	// Address book helpers
	GetAddrBook() AddrBook
	AddPeerToAddrBook(peer *NetworkPeer) error    // TODO(team): Not used yet
	RemovePeerToAddrBook(peer *NetworkPeer) error // TODO(team): Not used yet

	// This function was added to specifically support the RainTree implementation.
	// Handles the raw data received from the network and returns the data to be processed
	// by the application layer.
	HandleNetworkData(data []byte) ([]byte, error)
}
```

This handles general purpose bytes at the `Network` (i.e. not P2P module) interface layer, rather than an "envelope" Network message at the P2P module.

Lmk if this justifies the approach or not.

**Comment by Olshansk** [comment]
@derrandz 

**Legend for my response on review comments**
- Using `ACK` as neither agreement nor disagreement, but rather as "let's discuss later" for comments outside of the scope of this PR.
- Using `DONE` where changes were applied
- Using `AGGREGATE` when multiple comments span the same topic, but we still need to follow up on it afterwards
- Left some open comments inline where appropriate

**Some feedback review:**
- If something is a rename, and it spans multiple files, it’s better to leave it as a comment than a suggestion. Using Github’s “commit suggestion” feature would make it break.
- As much as I appreciate the positive feedback, leaving too many inline “I like this part” messages added noise to the review. I believe having one left at the top level is sufficient.

**Next steps and open questions remaining:**
1. If you feel I answered your comment (which I did not resolve), please either resolve it or leave a follow-up comment. If the follow-up comment is non-actionable, please resolve.
2. The reason for `NetworkSend` and `NetworkBroadcast` is because they are part of the `Network` interfaces and not the `P2PModules` Interface, which has a `Broadcast` and `Send` method. I did not want the two to be confused. Lmk what you think

File: README.md (hunk 13 above)
Comment (derrandz) [reply README.md:49]: Kudos for the suggestion, I will create an issue for this.

File: p2p/pre2p/module.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (derrandz) [reply p2p/pre2p/module.go:128]: Referencing your comment about how this avoids confusing the `Pre2PModule.Broadcast` with `Network.NetworkBroadcast`, I think I agree.

**Comment by derrandz** [comment]
@Olshansk Thanks for the feedback and for acting on the review.

I've resolved the parts I am okay with. Left open just one that has an actionable comment. There is one resolved conversation in particular in which I left a clarifying comment to make the case for renaming is [this](https://github.com/pokt-network/pocket/pull/80#discussion_r899209295).

Thank you.

File: p2p/pre2p/module.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (derrandz) [reply p2p/pre2p/module.go:28]: No. Of course. My bad, I gave a bad example by naming it `TCPTransport`. It should be just `Transport` for the reasons I mentioned in the [suggestion](https://github.com/pokt-network/pocket/pull/80#discussion_r895246150) I made before.

I would be open to keeping the name as is, but change it as a part of #86... Since really, per the implementation we have right now, it should keep the name `TransportLayerConn` as the interface defines `Conn`-like behavior like: `Read` and `Write`.

The name should change eventually though, to keep a faithful abstraction, as the `Conn` behavior and the `Transport` behavior will be separated.

File: p2p/pre2p/types/network.go (hunk 11 above)
Comment (Olshansk) [reply p2p/pre2p/types/network.go]: Renamed `TransportLayerConn to Transport` and pushed.

**Comment by Olshansk** [comment]
@derrandz @andrewnguyen22 I believe I've addressed all the comments, so please re-review/approve if everything looks good to merge.

**Comment by derrandz** [comment]
Hey @Olshansk I've created #93  for the changelog task and tagged it as good starter task for the community as we've discussed before in this [conversation](https://github.com/pokt-network/pocket/pull/80#discussion_r898622625).

Since that conversation is linked as an origin document to the issue, referencing this issue on the PR will leave good trace-ability.

**Review by andrewnguyen22 — APPROVED** [review]

**Review by derrandz — APPROVED** [review]
